File: packages/sdk-node/src/evidence.ts
/**
 * RuneSignal Node SDK — Evidence Bundles Resource
 * POST /api/v1/compliance/evidence-export
 * GET  /api/v1/compliance/evidence-bundles
 */

import { BaseClient } from './client';

export type Regulation = 'eu_ai_act' | 'iso_42001' | 'nist_ai_rmf' | 'hipaa' | 'sox' | 'gdpr' | 'pci_dss';
export type EvidenceFormat = 'json' | 'pdf' | 'zip';

export interface EvidencePreviewRequest {
  regulation: Regulation;
  dateFrom: string;   // ISO date YYYY-MM-DD
  dateTo: string;
  agentIds?: string[];
}

export interface EvidenceGap {
  article: string;
  description: string;
  severity: 'critical' | 'major' | 'minor';
  remediationHint: string;
}

export interface ArticleCoverage {
  article: string;
  title: string;
  covered: boolean;
  coveragePct: number;
  evidenceCount: number;
}

export interface EvidencePreview {
  regulation: Regulation;
  overallScore: number;
  articleCoverage: ArticleCoverage[];
  gaps: EvidenceGap[];
  periodFrom: string;
  periodTo: string;
  agentCount: number;
  totalEvents: number;
}

export interface GenerateEvidenceRequest {
  regulation: Regulation;
  dateFrom: string;
  dateTo: string;
  agentIds?: string[];
  includeAttachments?: boolean;
  signatoryEmail?: string;
}

export interface EvidenceBundle {
  id: string;
  tenantId: string;
  regulation: Regulation;
  status: 'generating' | 'ready' | 'failed';
  overallScore: number;
  articleCount: number;
  evidenceItemCount: number;
  signatureHash?: string;
  downloadUrl?: string;
  expiresAt?: string;
  createdAt: string;
}

export class EvidenceResource {
  constructor(private readonly client: BaseClient) {}

  /**
   * Dry-run preview — shows coverage score and gaps without persisting a bundle.
   *
   * @example
   * const preview = await client.evidence.preview({
   *   regulation: 'eu_ai_act',
   *   dateFrom: '2026-01-01',
   *   dateTo: '2026-03-31',
   * });
   * console.log(`Coverage: ${preview.overallScore}% · Gaps: ${preview.gaps.length}`);
   */
  async preview(request: EvidencePreviewRequest): Promise<EvidencePreview> {
    const query: Record<string, string> = {
      regulation: request.regulation,
      date_from: request.dateFrom,
      date_to: request.dateTo,
    };
    if (request.agentIds?.length) {
      query.agent_ids = request.agentIds.join(',');
    }

    const raw: any = await (this.client as any).request(
      'GET',
      '/api/v1/compliance/evidence-preview',
      { query }
    );

    return {
      regulation: raw.regulation,
      overallScore: raw.overall_score,
      articleCoverage: (raw.article_coverage || []).map((a: any) => ({
        article: a.article,
        title: a.title,
        covered: a.covered,
        coveragePct: a.coverage_pct,
        evidenceCount: a.evidence_count,
      })),
      gaps: (raw.gaps || []).map((g: any) => ({
        article: g.article,
        description: g.description,
        severity: g.severity,
        remediationHint: g.remediation_hint,
      })),
      periodFrom: raw.period_from,
      periodTo: raw.period_to,
      agentCount: raw.agent_count,
      totalEvents: raw.total_events,
    };
  }

  /**
   * Generate and persist a signed evidence bundle.
   * The bundle is cryptographically signed (Ed25519) and stored for 10 years.
   *
   * @example
   * const bundle = await client.evidence.generate({
   *   regulation: 'eu_ai_act',
   *   dateFrom: '2026-01-01',
   *   dateTo: '2026-03-31',
   * });
   */
  async generate(request: GenerateEvidenceRequest): Promise<EvidenceBundle> {
    const raw: any = await (this.client as any).request(
      'POST',
      '/api/v1/compliance/evidence-export',
      {
        body: {
          regulation: request.regulation,
          date_from: request.dateFrom,
          date_to: request.dateTo,
          agent_ids: request.agentIds,
          include_attachments: request.includeAttachments,
          signatory_email: request.signatoryEmail,
        },
      }
    );

    return mapBundle(raw);
  }

  /** Download a bundle in the specified format. Returns the raw body. */
  async download(bundleId: string, format: EvidenceFormat = 'json'): Promise<unknown> {
    return (this.client as any).request('GET', `/api/v1/compliance/evidence-bundles/${bundleId}`, {
      query: { format },
    });
  }

  /** List all previously generated evidence bundles. */
  async list(options: { regulation?: Regulation; limit?: number } = {}): Promise<EvidenceBundle[]> {
    const query: Record<string, string> = {};
    if (options.regulation) query.regulation = options.regulation;
    if (options.limit)      query.limit      = String(options.limit);

    const raw: any[] = await (this.client as any).request(
      'GET', '/api/v1/compliance/evidence-bundles', { query }
    );
    return raw.map(mapBundle);
  }
}

function mapBundle(raw: any): EvidenceBundle {
  return {
    id:                raw.id,
    tenantId:          raw.tenant_id,
    regulation:        raw.regulation,
    status:            raw.status,
    overallScore:      raw.overall_score,
    articleCount:      raw.article_count,
    evidenceItemCount: raw.evidence_item_count,
    signatureHash:     raw.signature_hash,
    downloadUrl:       raw.download_url,
    expiresAt:         raw.expires_at,
    createdAt:         raw.created_at,
  };
}

File: packages/sdk-python/runesignal/evidence.py
"""RuneSignal Python SDK — Evidence Bundles"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional

from .http import BaseHTTPClient

Regulation = Literal["eu_ai_act", "iso_42001", "nist_ai_rmf", "hipaa", "sox", "gdpr", "pci_dss"]
EvidenceFormat = Literal["json", "pdf", "zip"]


@dataclass
class EvidenceGap:
    article: str
    description: str
    severity: str
    remediation_hint: str


@dataclass
class ArticleCoverage:
    article: str
    title: str
    covered: bool
    coverage_pct: float
    evidence_count: int


@dataclass
class EvidencePreview:
    regulation: str
    overall_score: float
    article_coverage: List[ArticleCoverage]
    gaps: List[EvidenceGap]
    period_from: str
    period_to: str
    agent_count: int
    total_events: int


@dataclass
class EvidenceBundle:
    id: str
    tenant_id: str
    regulation: str
    status: str
    overall_score: float
    article_count: int
    evidence_item_count: int
    created_at: str
    signature_hash: Optional[str] = None
    download_url: Optional[str] = None
    expires_at: Optional[str] = None


class EvidenceResource:
    def __init__(self, http: BaseHTTPClient):
        self._http = http

    async def preview(
        self,
        regulation: str,
        date_from: str,
        date_to: str,
        agent_ids: Optional[List[str]] = None,
    ) -> EvidencePreview:
        query: Dict[str, str] = {
            "regulation": regulation,
            "date_from": date_from,
            "date_to": date_to,
        }
        if agent_ids:
            query["agent_ids"] = ",".join(agent_ids)

        raw = await self._http.request("GET", "/api/v1/compliance/evidence-preview", query=query)
        return EvidencePreview(
            regulation=raw["regulation"],
            overall_score=raw["overall_score"],
            article_coverage=[
                ArticleCoverage(
                    article=a["article"],
                    title=a["title"],
                    covered=a["covered"],
                    coverage_pct=a["coverage_pct"],
                    evidence_count=a["evidence_count"],
                )
                for a in raw.get("article_coverage", [])
            ],
            gaps=[
                EvidenceGap(
                    article=g["article"],
                    description=g["description"],
                    severity=g["severity"],
                    remediation_hint=g["remediation_hint"],
                )
                for g in raw.get("gaps", [])
            ],
            period_from=raw["period_from"],
            period_to=raw["period_to"],
            agent_count=raw.get("agent_count", 0),
            total_events=raw.get("total_events", 0),
        )

    async def generate(
        self,
        regulation: str,
        date_from: str,
        date_to: str,
        agent_ids: Optional[List[str]] = None,
        include_attachments: bool = False,
        signatory_email: Optional[str] = None,
    ) -> EvidenceBundle:
        raw = await self._http.request(
            "POST",
            "/api/v1/compliance/evidence-export",
            body={
                "regulation": regulation,
                "date_from": date_from,
                "date_to": date_to,
                "agent_ids": agent_ids,
                "include_attachments": include_attachments,
                "signatory_email": signatory_email,
            },
        )
        return _map_bundle(raw)

    async def download(self, bundle_id: str, fmt: EvidenceFormat = "json") -> Any:
        return await self._http.request(
            "GET",
            f"/api/v1/compliance/evidence-bundles/{bundle_id}",
            query={"format": fmt},
        )

    async def list(
        self,
        regulation: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[EvidenceBundle]:
        query: Dict[str, str] = {}
        if regulation: query["regulation"] = regulation
        if limit:      query["limit"]      = str(limit)
        raw_list = await self._http.request("GET", "/api/v1/compliance/evidence-bundles", query=query)
        return [_map_bundle(r) for r in raw_list]


def _map_bundle(raw: Dict[str, Any]) -> EvidenceBundle:
    return EvidenceBundle(
        id=raw["id"],
        tenant_id=raw.get("tenant_id", ""),
        regulation=raw["regulation"],
        status=raw.get("status", "ready"),
        overall_score=raw.get("overall_score", 0.0),
        article_count=raw.get("article_count", 0),
        evidence_item_count=raw.get("evidence_item_count", 0),
        signature_hash=raw.get("signature_hash"),
        download_url=raw.get("download_url"),
        expires_at=raw.get("expires_at"),
        created_at=raw.get("created_at", ""),
    )
